### ISSUE TITLE
Run some workflows based on which files have changed

### ISSUE THREAD

**Issue by stoyanK7** [checkstyle/checkstyle#12936]
Why not limit some workflows based on the files that have changed in the PR? I don't think it is necessary to run `pitest`(for example) if I have only made changes in the `.ci` folder.

Github workflows allows this: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

> When using the `push` and `pull_request` events, you can configure a workflow to run based on what file paths are changed.

**Comment by nrmancuso** [comment]
Before we can approve this issue, we need a detailed plan, and may require separate issues for each. I have often thought about doing the same, but I am not sure if the lift is worth it. We do not pay for any CI, and have rarely hit any walls with usage, so ROI in my opinion is low.

**Comment by stoyanK7** [comment]
Makes sense. I'd say limit at least the bigger ones like `checker` and `pitest` that take 20-30(?)% of the whole CI usage.

**Comment by Vyom-Yadav** [comment]
I have also often thought about implementing this. Yes, we don't pay for CI but a lot of times many actions are in the queue because there is no runner available. For eg see https://github.com/checkstyle/checkstyle/actions/runs/4571603449, simple action to generate a site is in the queue for 16 minutes (as of now).

Limiting heavy actions like pitest, checker, etc. will free some runners so that all other actions don't have to wait for a long time to be picked up by a runner.

**Comment by Vyom-Yadav** [comment]
Plus many jobs get canceled too:
> Canceling since a higher priority waiting request for 'Site-refs/heads/master' exists

**Comment by nrmancuso** [comment]
How do we handle the case where we change some script in `.ci` that runs checker/pitest/etc? :)

Again, we need to take this on a case by case basis, please provide a concrete example of "We will not run x if y, z"

**Comment by romani** [comment]
we had simple form of skips CI before, it endup in numerous cases that were unfortunately skipped, and eventually we end up removing skips at all, as jobs migrated a lot between CIs.
Idea is good, but it requires thorough attention to details.

**Comment by stoyanK7** [comment]
Putting this here as an example. If it gets approval, we can move to a separate issue.

Let's take `pitest`.

```
$ git grep pitest | sed "s/\([^<>]\):.*/\1/" | sort -u
cdg-pitest-licence.txt
.ci/pitest.sh
.ci/pitest-survival-check-xml.groovy
.circleci/config.yml
.ci/validation.sh
config/checkstyle-non-main-files-suppressions.xml
config/checkstyle-resources-suppressions.xml
config/intellij-idea-inspection-scope.xml
config/intellij-idea-inspections.properties
config/jsoref-spellchecker/exclude.pl
config/jsoref-spellchecker/whitelist.words
config/org.eclipse.jdt.core.prefs
config/pitest-suppressions/pitest-ant-suppressions.xml
config/pitest-suppressions/pitest-api-suppressions.xml
config/pitest-suppressions/pitest-coding-1-suppressions.xml
config/pitest-suppressions/pitest-coding-2-suppressions.xml
config/pitest-suppressions/pitest-common-2-suppressions.xml
config/pitest-suppressions/pitest-common-suppressions.xml
config/pitest-suppressions/pitest-filters-suppressions.xml
config/pitest-suppressions/pitest-header-suppressions.xml
config/pitest-suppressions/pitest-imports-suppressions.xml
config/pitest-suppressions/pitest-indentation-suppressions.xml
config/pitest-suppressions/pitest-javadoc-suppressions.xml
config/pitest-suppressions/pitest-main-suppressions.xml
config/pitest-suppressions/pitest-misc-suppressions.xml
config/pitest-suppressions/pitest-packagenamesloader-suppressions.xml
config/pitest-suppressions/pitest-regexp-suppressions.xml
config/pitest-suppressions/pitest-tree-walker-suppressions.xml
config/pitest-suppressions/pitest-utils-suppressions.xml
config/pitest-suppressions/pitest-xpath-suppressions.xml
.github/workflows/pitest.yml
pom.xml
README.md
.semaphore/semaphore.yml
src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
src/test/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheckTest.java
src/test/java/com/puppycrawl/tools/checkstyle/checks/sizes/RecordComponentNumberCheckTest.java
src/test/java/com/puppycrawl/tools/checkstyle/JavaAstVisitorTest.java
src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java
src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/InputGenericWhitespaceAtStartOfTheLine.java
src/xdocs/releasenotes_old_6-0_7-8.xml
src/xdocs/releasenotes_old_8-0_8-34.xml
src/xdocs/releasenotes.xml

```

A list of patterns that should run `pitest` based on the output above:

- `.circleci/config.yml`
- `**pitest**`
- `.ci/validation.sh`
- `config/checkstyle-non-main-files-suppressions.xml`
- `config/checkstyle-resources-suppressions.xml`
- `config/intellij-idea-inspection-scope.xml`
- `config/intellij-idea-inspections.properties`
- `config/jsoref-spellchecker/exclude.pl`
- `config/jsoref-spellchecker/whitelist.words`
- `config/org.eclipse.jdt.core.prefs`
- `pom.xml`
- `README.md` 
- `.semaphore/semaphore.yml`
- `src/**`

**Comment by romani** [comment]
I think better to do pattern of skip.
We will never remember to update dependency list, so we will introduce new dependency to code and pitest will be skipped, it will be hard to notice it.

What if we focus on ignore list, what updates to skip. So we know what patterns to remove from diff status output.

**Comment by nrmancuso** [comment]
>I think better to do pattern of skip.

Yes, we must focus on exclusion only, but we also need to do some research and find examples of PRs where we didn't need pitest to run. We could also think about commit message exclusion, and even add new prefix like "skip_ci: ...`
